### PR TITLE
Integrate te permutation

### LINKED ISSUE #2585: [JAX] Integrate Permutation to MaxText
After permutation primitive is merged, needs to integrate to MaxText for full model testing

### PR DESCRIPTION
**Pull request by jberchtold-nvidia**
# Description

This integrates TE's permutation to maxtext, offering an alternative to the pure jax implementation of permutation in the MoE layer.

1. This should follow the same flow as the previous pure-jax implementation with global permute -> ragged A2A / ring of experts -> local permute. With global_permute calling the TE's _permute_kernel, and local_permute calling sort_chunks_by_index. Similarly for unpermute.

2. This TE's permutation integration also offers support for padding each expert with zero-ed tokens up to an alignment, controlled via te_permutation_align_size, which is set to 128 by default. This approach differs from maxtext's original approach of padding at the end of the whole input tensor, by instead padding per-expert.

If the change fixes a bug or a Github issue, please include a link, e.g.,:
fixes https://github.com/NVIDIA/TransformerEngine/issues/2585

# Tests

I run the mixtral-8x7b model (with an adjusted number of decoder later to 1-4 (from 32) on a node on ptyche system (GB200),  with use_te_permutation set to True in base.yaml (or command line), and padding set to 128 (by default).

I still observe the same convergence loss value as the maxtext's original implementation after the same number of steps. 
Here is a table of results when num encoder layer is set to 4, and there is 30 steps

Config | Status | Final Loss
-- | -- | --
ep1_fsdp2_tefalse_sparsefalse_align0_ringfalse | SUCCESS | 5.1460
ep1_fsdp2_tefalse_sparsefalse_align0_ringtrue | SUCCESS | 5.1460
ep1_fsdp2_tefalse_sparsetrue_align0_ringfalse | SUCCESS | 5.0940
ep1_fsdp2_tefalse_sparsetrue_align0_ringtrue | SUCCESS | 5.0770
ep1_fsdp2_tetrue_sparsefalse_align0_ringfalse | SUCCESS | 5.1460
ep1_fsdp2_tetrue_sparsefalse_align0_ringtrue | SUCCESS | 5.1460
ep1_fsdp2_tetrue_sparsefalse_align128_ringfalse | SUCCESS | 5.1510
ep1_fsdp2_tetrue_sparsefalse_align128_ringtrue | SUCCESS | 5.1470
ep1_fsdp2_tetrue_sparsetrue_align0_ringfalse | SUCCESS | 5.0770
ep1_fsdp2_tetrue_sparsetrue_align0_ringtrue | SUCCESS | 5.0790
ep1_fsdp2_tetrue_sparsetrue_align128_ringfalse | SUCCESS | 5.0780
ep1_fsdp2_tetrue_sparsetrue_align128_ringtrue | SUCCESS | 5.0880
ep2_fsdp2_tefalse_sparsefalse_align0_ringfalse | SUCCESS | 6.5270
ep2_fsdp2_tefalse_sparsefalse_align0_ringtrue | SUCCESS | 6.5270
ep2_fsdp2_tefalse_sparsetrue_align0_ringfalse | SUCCESS | 6.4840
ep2_fsdp2_tefalse_sparsetrue_align0_ringtrue | SUCCESS | 6.4810
ep2_fsdp2_tetrue_sparsefalse_align0_ringfalse | SUCCESS | 6.5270
ep2_fsdp2_tetrue_sparsefalse_align0_ringtrue | SUCCESS | 6.5270
ep2_fsdp2_tetrue_sparsefalse_align128_ringfalse | SUCCESS | 6.5280
ep2_fsdp2_tetrue_sparsefalse_align128_ringtrue | SUCCESS | 6.5270
ep2_fsdp2_tetrue_sparsetrue_align0_ringfalse | SUCCESS | 6.4810
ep2_fsdp2_tetrue_sparsetrue_align0_ringtrue | SUCCESS | 6.4930
ep2_fsdp2_tetrue_sparsetrue_align128_ringfalse | SUCCESS | 6.5990
ep2_fsdp2_tetrue_sparsetrue_align128_ringtrue | SUCCESS | 6.4810

# Checklist

Before submitting this PR, please make sure (put X in square brackets):
- [x] I have performed a self-review of my code. For an optional AI review, add the `gemini-review` label.
- [x] I have necessary comments in my code, particularly in hard-to-understand areas.
- [x] I have run end-to-end tests tests and provided workload links above if applicable.
- [x] I have made or will make corresponding changes to the doc if needed, including adding new documentation pages to the relevant Table of Contents (toctree directive) as explained in [our documentation](https://maxtext.readthedocs.io/en/latest/development.html#adding-new-documentation-files).

### CODE REVIEW

File: src/MaxText/configs/base.yml [hunk 1]
```diff
@@ -181,6 +182,8 @@ load_balance_loss_weight: 0.0 # weight for the load balance loss
 use_random_routing: false # whether to use random routing for debug/test purpose
 use_custom_sort_vjp: true # whether to use a custom VJP sort for efficient backward pass processing in sparse matmul
 use_ring_of_experts: false # whether to use ring of experts for sparse matmul expert parallelism
+use_te_permutation: false # whether to use TransformerEngine permutation kernels for MoE token dispatch/combine
```
Comment (jberchtold-nvidia) [inline src/MaxText/configs/base.yml]: nit SR: might  be nice to prefix all our TE fields with a common prefix like  `te_` or  `transformerengine_`. We don't have a convention currently since our existing functionality is just a new enum value of existing fields, but we haven't added new fields

File: src/MaxText/layers/te_permutation.py [hunk 2]
```diff
@@ -0,0 +1,354 @@
...
+  """
+  # Flatten to 2D if needed: [num_tokens, num_experts_per_tok]
+  original_shape = top_k_indices.shape
+  if top_k_indices.ndim == 3:
+    num_tokens = original_shape[0] * original_shape[1]
+    top_k_indices = top_k_indices.reshape(num_tokens, -1)
+
+  # Create one-hot encoding for each selected expert
+  # Shape: [num_tokens, num_experts_per_tok, num_experts]
+  one_hot = jax.nn.one_hot(top_k_indices, num_classes=num_experts, dtype=jnp.int32)
```
Comment (jberchtold-nvidia) [inline src/MaxText/layers/te_permutation.py:89]: Is most of the JAX logic here to convert the routing map to a version TE's permute can use? And a lot of this logic go away once we integrate the TE router?

When possible, I like to keep complexity in TE itself cause we can merge changes faster than upstreaming to MaxText and others can re-use the functionality directly from TE. But I agree things like checking if TE methods can be imported would be useful to have in MaxText itself like you've done here, especially since there are many instances where permute is called in moe.py.

So if this is temporary complexity until we integrate the router, then I'm okay  with this as is. But if the JAX operations here are permanent complexity, then we  should consider moving it  inside TE and only exposing a minimal interface for it.

File: src/MaxText/layers/moe.py [hunk 3]
```diff
@@ -769,6 +950,136 @@ def local_permute(
...
+    # Target: [(s0,e0), (s1,e0), ..., (s0,e1), (s1,e1), ...]
+    indices_matrix = jnp.arange(num_expert_shards * local_expert_size).reshape(
+        num_expert_shards, local_expert_size
+    )
+    sorted_chunk_indices = indices_matrix.T.reshape(-1)
+
+    # Use TE sort_chunks_by_index
+    # NOTE: With the kernel fix, tokens beyond sum(split_sizes) map to themselves
+    # (identity mapping) to avoid corrupting valid data.
+    sorted_inputs, sort_map = te_permutation.te_sort_chunks_by_expert(
```
Comment (jberchtold-nvidia) [inline src/MaxText/layers/moe.py:1191]: Same question as in te_permutation.py for whether these  JAX operations are permanent complexity or temporary conversions until we integrate the router

File: src/MaxText/layers/moe.py [hunk 4]
```diff
@@ -892,7 +1203,7 @@ def gmm(inputs, kernel, tiling, group_sizes, expert_assignments, weight_gather_a
       kernel = kernel.astype(self.dtype)
 
       lhs_quantize_dtype, rhs_quantize_dtype = None, None
-      if self.quant is not None:
+      if self.quant is not None and hasattr(self.quant, 'quant_dg'):
```
Comment (jberchtold-nvidia) [inline src/MaxText/layers/moe.py:1377]: Good fix, thanks!

File: src/MaxText/layers/moe.py [hunk 5]
```diff
@@ -1151,24 +1539,80 @@ def wrapper(x, logits, pre_bias_logits, w0, w1, wo, w0_bias, w1_bias, wo_bias, r
...
+                  x,
+                  te_all_shards_tokens_per_expert,
+                  local_expert_size,
+                  expert_shard_id,
+                  num_expert_parallelism,
+              )
+              local_sorted_indices = None  # Not used in TE path
+            else:
+              # For MT local permute, we need UNPADDED group sizes
+              # In mixed mode (TE global + MT local), use te_original_group_sizes
```
Comment (jberchtold-nvidia) [inline src/MaxText/layers/moe.py]: The existing logic is already pretty dense with lots of if/else branches, so I'm hesitant to add more complexity without breaking it out into helper functions to help with our own understanding but also to make sure we don't break existing communication logic while rebasing and maintaining our fork.

What are the barriers to having a shared interface between TE permute and JAX permute (and same for local_permute)? Ideally we'd just call `self.permute` like before and internally that  would either call TE or JAX. For the TE case, the conversion of the routing map could be done internally to self.permute.

I see this "mixed mode (TE global + MT local)" breaks this shared interface because it requires additional variables like `te_original_group_sizes`. Seems to be the same above for  "This is needed by TE local permute even when using MT global"

Are these mixed mode cases the only ones that breaks a shared interface or are there more issues that prevent a shared interface here? I see `te_pad_offsets` and other additional outputs that may pose issues for the shared interface

File: src/MaxText/configs/base.yml (hunk 1 above)
Comment (tdophung) [reply src/MaxText/configs/base.yml]: I can make this one "te_permutation_impl" = True/False. Then we can set a pattern for new fields to be te_*

File: src/MaxText/layers/te_permutation.py (hunk 2 above)
Comment (tdophung) [reply src/MaxText/layers/te_permutation.py:89]: This part is temporary complexity because we have not integrated router. I will try to see what else that are not temporary, and move them to TE if exist

Comment (jberchtold-nvidia) [reply src/MaxText/layers/te_permutation.py:89]: Sounds good, thanks! I think the most important parts to refactor would be anything in moe.py that will cause merge conflicts. The other changes can wait until the router is integrated and the code is in a more final state. Merging into our MaxText fork is equally as easy as merging to TE  (unless we have MaxText merge conflicts), so the organization of which complexity can be hidden in TE only really becomes an issue when we try to officially upstream and maintain this.

File: src/MaxText/layers/moe.py (hunk 3 above)
Comment (tdophung) [reply src/MaxText/layers/moe.py:1191]: these are permanent ugliness because it is to calculate the inputs (split sizez, sorted_chunk_indices) for TE's sprt_chunk_by_experts, from the ragged a2a params, which will not be available inside TE, but only on maxtext side

File: src/MaxText/layers/te_permutation.py (hunk 2 above)
Comment (tdophung) [reply src/MaxText/layers/te_permutation.py:89]: I did not hide everything in TE as some  lines are only there because of ragged a2a ( the output of this operation is not excatly what TE's input is, so some tranformation needs to happen), so I instead wrap it inside of the common permute() API used for both maxtext and TE. This helps reduce the amount of if-else in the main wrapper() body, and hide it under permute(). local_permute(), unpermute(), local_unpermute(), which would help with merging to upstream

Comment (jberchtold-nvidia) [reply src/MaxText/layers/te_permutation.py:89]: Perfect, thanks!

**Comment by jberchtold-nvidia** [comment]
LGTM!

**Review by tdophung — APPROVED** [review]
Approve on behalf of Jeremy
